### ISSUE TITLE
failing C++14 compilation for mxnet 1.5.1

### ISSUE THREAD

**Issue by olk** [apache/incubator-mxnet#17156]
## Description
mxnet compiled with USE_CXX14_IF_AVAILABLE=ON forced compilation error in DMLC.

### Error Message
```
/home/graemer/.cache/yay/mxnet-git/src/mxnet-git/include/dmlc/thread_group.h:77:26: Fehler: »shared_timed_mutex« in Namensbereich »std« bezeichnet keinen Typ
   77 | using SharedMutex = c;
/home/graemer/.cache/yay/mxnet-git/src/mxnet-git/include/dmlc/thread_group.h:77:21: Anmerkung: »std::shared_timed_mutex« ist erst ab C++14 verfügbar
   77 | using SharedMutex = std::shared_timed_mutex;
```

I guess that -std=C++14 isn't correctly passed to DMLC because file thread_group.h includes C++14 features (like std::shared_timed_mutex) only if 
DMLC_USE_CXX14 is defined or __cplusplus > 201103L

I think that __cplusplus > 201103L is misleading because only 201402L does identify complete C++14 (e.g. std::shared_timed_mutex -> https://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros#C.2B.2B14).

Please note that gcc-9.2.0 uses C++14 per default.

## Environment

We recommend using our script for collecting the diagnositc information. Run the following command and paste the outputs below:
```----------Python Info----------
Version      : 3.8.0
Compiler     : GCC 9.2.0
Build        : ('default', 'Oct 23 2019 18:51:26')
Arch         : ('64bit', 'ELF')
------------Pip Info-----------
Version      : 19.2.3
Directory    : /usr/lib/python3.8/site-packages/pip
----------MXNet Info-----------
Version      : 1.6.0
Directory    : /usr/lib/python3.8/site-packages/mxnet
Num GPUs     : 2
Hashtag not found. Not installed from pre-built package.
----------System Info----------
Platform     : Linux-5.4.3-arch1-1-x86_64-with-glibc2.2.5
system       : Linux
node         : e5lx
release      : 5.4.3-arch1-1
version      : #1 SMP PREEMPT Fri, 13 Dec 2019 09:39:02 +0000

```

**Comment by olk** [comment]
I notice that file mxnet/CMakeLists.txt sets -DDMLC_USE_CXX11=1 and -DDMLC_USE_CXX14=1 if SUPPORT_CXX14 is valid.
Shouldn't only -DDMLC_USE_CXX14=1 be specified?!

**Comment by leezu** [comment]
Are you using the cmake or Makefile based compilation?

**Comment by olk** [comment]
cmake

**Comment by samskalicky** [comment]
This is not a supported build flow, we havent tried nor do we regularly build/test building with c++14. 
@mxnet-label-bot update [Cmake, Feature request]

**Comment by leezu** [comment]
@samskalicky why not? We have a `USE_CXX14_IF_AVAILABLE` option:

https://github.com/apache/incubator-mxnet/blob/44784c2554f0330b94d8bde411394da89405c397/CMakeLists.txt#L58

**Comment by olk** [comment]
Seams to be fixed for current code (1.6.0 rc2).
